<p>
  <img src="https://upload.wikimedia.org/wikipedia/commons/f/fb/Escudo-UdeA.svg"
       alt="UdeA logo"
       height="190px"
       align="left"
       hspace="15px">
</p>

<h1 style="margin-top: 20px;"><b>2. Uncertainty report</b></h1>

<hr>

<div align="right" style="font-size: 14px; line-height: 1.4;">
  <b>Estiven Castrillon Alzate</b><br>
  Institute of Physics<br>
  <i>Universidad de Antioquia</i><br>
  <span style="font-size: 12px;">Date: TO BE DEFINED</span>
</div>

<br><br>

# Table of Contents
1. [Introduction](#introduction)
2. [Theory](#theory)
3. [Analysis and code](#analysis-and-code)
    - [Analysis of pendulum properties](#analysis-of-pendulum-properties)
    - [Determinantion of G](#determinantion-of-g)
4. [Results](#results)
5. [Conclusions](#conclusions)
6. [References](#references)

# Introduction

The precise measurement and analysis of physical quantities are fundamental aspects of experimental physics, as they allow us to understand the underlying principles governing natural phenomena. In this report, we focus on the study of uncertainties associated with the measurement of the period of a pendulum and the subsequent determination of the local gravitational acceleration. The experimental method involved measuring the length of a simple pendulum and recording the time taken for multiple oscillations using a stopwatch. Several sets of measurements were taken for different numbers of oscillations (1, 5, 7, 9, and 10), and the period was calculated by dividing the total time by the number of oscillations in each case, all the data were taken at Universidad de Antioquia in Medellín, Colombia. By performing repeated measurements and applying statistical methods, we aim to quantify both random and systematic errors, thereby improving the reliability of our experimental results. This approach not only enhances our comprehension of measurement techniques but also underscores the importance of uncertainty analysis in scientific experimentation. The idea is also to compare the data extracted from the analysis, specifically from the calculus of the gravitational acceleration, and compare it with the current best approximation of the gravity in Medellín, Antioquia, Colombia done by Servicio Geológico Colombiano (Estaciones Red de Gravedad Absoluta Para Colombia -RGAC-, 2024).

# Theory

The period of a simple pendulum is governed by the relationship $$T = 2\pi \sqrt{\frac{l}{g}}$$ where $T$  is the period, $l$ is the length of the pendulum, and $g$ is the acceleration due to gravity. Accurate determination of $g$ requires careful measurement of both $l$ and $T$, as well as a thorough analysis of the uncertainties involved. Uncertainties can arise from instrumental limitations, human reaction time, and random fluctuations in repeated measurements. The propagation of these uncertainties is essential for reporting meaningful results and for comparing experimental findings with theoretical predictions. The methodology and significance of uncertainty analysis in experimental physics are well documented in the literature, such as in the "Notas de Física Experimental" by Edgar Rueda (Notas de Física Experimental — Notas de Física Experimental, 2022) as well as in the book by professor Jorge Mahecha (TratamientoEstadísticoDeDatosExperimentales, 2009)


# Analysis and code

## Analysis of pendulum properties

### Measurement of the pendulum length

Is it worthwhile to perform repeated measurements of the variable, or is a single measurement sufficient? For practical purposes of the activity, multiple measurements of the pendulum length were not taken. However, ideally this should be done in order to produce a more complete uncertainty report for the estimated gravitational acceleration obtained from the period and length data of the pendulum. This is due to the dependence of gravity on both the pendulum length and its period, as shown in the following process:

$$
T = 2\pi \sqrt{\frac{l}{g}} \tag{1}
$$

where $T$ is the pendulum period, $g$ is the acceleration due to gravity, and $l$ is the length of the pendulum. Solving for $g$, we obtain:

$$
g = \frac{4\pi^2 l}{T^2} \tag{2}
$$

The length of the pendulum is given by:

$$
l = (73.6 \pm 0.1)\,\text{cm}
$$


### Measurement of the pendulum oscillation period

It is preferable to measure the time of several periods and then divide. Why?

Since the oscillation angle is very small and the length of the string is not very large, the oscillations are quite fast. This can be counterproductive when measuring time with a stopwatch. By timing several periods instead of just one, the probability of error when starting and stopping the stopwatch is reduced, leading to a more accurate estimate of the pendulum period.

In [1]:
import pandas as pd
import numpy as np

from math import radians

Importación de datos

In [2]:
# # # Data import using files from the local machine
# oscillation_data = pd.read_excel('2.0-pendulum_oscillation_data.xlsx')

# Data import based on a JSON file
# This approach does not require having the data file on the local machine
oscillation_data_json = '{"1_oscilacion(s)":{"0":1.58,"1":1.65,"2":1.54,"3":1.63,"4":1.61},"5_oscilaciones(s)":{"0":8.73,"1":8.77,"2":8.84,"3":8.81,"4":8.81},"7_oscilaciones(s)":{"0":12.21,"1":12.25,"2":12.24,"3":12.26,"4":12.16},"9_oscilaciones(s)":{"0":15.82,"1":15.98,"2":15.92,"3":15.93,"4":15.9},"10_oscilaciones(s)":{"0":17.69,"1":17.46,"2":17.6,"3":17.56,"4":17.57}}'

oscillation_data = pd.read_json(oscillation_data_json)
oscillation_data

/tmp/ipykernel_10927/1386846615.py:8: FutureWarning: Passing literal json to 'read_json' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  oscillation_data = pd.read_json(oscillation_data_json)


1_oscilacion(s)  5_oscilaciones(s)  7_oscilaciones(s)  9_oscilaciones(s)  \
0             1.58               8.73              12.21              15.82   
1             1.65               8.77              12.25              15.98   
2             1.54               8.84              12.24              15.92   
3             1.63               8.81              12.26              15.93   
4             1.61               8.81              12.16              15.90   

   10_oscilaciones(s)  
0               17.69  
1               17.46  
2               17.60  
3               17.56  
4               17.57

Dividing each value in the column by the number of oscillations
in order to obtain the oscillation period


In [3]:
oscillation_periods = oscillation_data.copy()  # Creating a copy of the original data

for column in oscillation_periods:
    number_of_oscillations = int(column.split('_')[0])
    oscillation_periods[column] = round(
        oscillation_periods[column] / number_of_oscillations, 2
    )

oscillation_periods

1_oscilacion(s)  5_oscilaciones(s)  7_oscilaciones(s)  9_oscilaciones(s)  \
0             1.58               1.75               1.74               1.76   
1             1.65               1.75               1.75               1.78   
2             1.54               1.77               1.75               1.77   
3             1.63               1.76               1.75               1.77   
4             1.61               1.76               1.74               1.77   

   10_oscilaciones(s)  
0                1.77  
1                1.75  
2                1.76  
3                1.76  
4                1.76

In [4]:
print('-----Average of each column-----')
average_periods = round(oscillation_periods.mean(), 2)
average_periods

-----Average of each column-----


1_oscilacion(s)       1.60
5_oscilaciones(s)     1.76
7_oscilaciones(s)     1.75
9_oscilaciones(s)     1.77
10_oscilaciones(s)    1.76
dtype: float64

Similarly, discuss whether it is worthwhile to perform repeated measurements of the variable or if a single measurement is sufficient. What strategy should be used to reduce the error caused by manual operation of the stopwatch?

* Performing repeated measurements is crucial to assess the consistency and precision of the results. Through repeated measurements, it is possible to calculate averages and standard deviations, which provide a more accurate idea of the magnitude of the uncertainty associated with the measurements. The more repeated measurements are performed, the better the estimation of the true magnitude of the variable and its uncertainty.

* First, the same person should measure the time while setting the pendulum to a 15-degree displacement. This minimizes errors associated with the reaction time of the person when starting the stopwatch and releasing the pendulum. Second, optimize the observer's line of sight relative to the pendulum to avoid starting or stopping the timer too early or late due to a poor perspective. Third, ensure the best possible environmental conditions: have a space with minimal air currents to avoid wind friction affecting the measurements, and use a stable surface that is not disturbed by the pendulum's motion, since any movement of the surface would alter the results.


Report T properly with its respective uncertainty. Consider that you must estimate a total uncertainty that accounts for both the instrument’s precision and the uncertainty due to random fluctuations.

Construct a table that contains, for a given number of measurements, columns with the expected value, the standard deviation, the square of the standard uncertainty, the square of half the instrument resolution, and the quadrature sum of the standard uncertainty and the instrument resolution. Do this for different numbers of measurements. Comment on the behavior of the values in the different columns.


Report of T (period)

In [5]:
# Sample standard deviation
periods_array = oscillation_periods.values.flatten()
std_dev = periods_array.std()
mean_period = periods_array.mean()

# Standard uncertainty of the mean
std_uncertainty = std_dev / np.sqrt(len(periods_array))

# Measurement instrument uncertainty (stopwatch)
stopwatch_uncertainty = 0.01

# Total uncertainty
total_uncertainty = np.sqrt((std_uncertainty ** 2) + (stopwatch_uncertainty ** 2))

# Report of T with uncertainty
print(f'T_period = ({round(mean_period, 2)} ± {round(total_uncertainty, 2)}) seconds')

T_period = (1.73 ± 0.02) seconds


Statistical samples table

In [6]:
periods_array = oscillation_periods.values.flatten()

def statistical_table_samples(sample_sizes):
    """
    Calculates statistical measures for a sample of data.

    Parameters:
        sample_sizes (list): List of sample sizes to consider.

    Returns:
        pd.DataFrame: A DataFrame with statistical measures for each sample size.
    """
    instrument_uncertainty = 0.01

    results = []
    for sample_size in sample_sizes:
        # Taking the sample from the data
        sample_data = periods_array[:sample_size]

        mean_data = np.round(sample_data.mean(), 2)
        std_dev_data = np.round(sample_data.std(), 2)

        # Standard uncertainty squared
        std_uncert_sq = np.round((std_dev_data / np.sqrt(sample_size)) ** 2, 4)

        # Square of half the instrument resolution
        half_instrument_sq = np.round((instrument_uncertainty / 2) ** 2, 5)

        # Quadrature sum of the standard uncertainty and instrument resolution
        total_uncertainty_sample = np.round(
            np.sqrt(std_uncert_sq + (instrument_uncertainty ** 2)), 2
        )

        # Defining the table with statistical measures
        results.append({
            'Sample Size': sample_size,
            'Mean': mean_data,
            'Standard Deviation': std_dev_data,
            'Standard Uncertainty Squared': std_uncert_sq,
            'Half Instrument Uncertainty Squared': half_instrument_sq,
            'Total Uncertainty': total_uncertainty_sample
        })

    df_results = pd.DataFrame(results)
    return df_results

In [7]:
statistical_table_samples(sample_sizes=[5, 10, 15, 25, 30])

Sample Size  Mean  Standard Deviation  Standard Uncertainty Squared  \
0            5  1.72                0.07                        0.0010   
1           10  1.73                0.06                        0.0004   
2           15  1.72                0.07                        0.0003   
3           25  1.73                0.07                        0.0002   
4           30  1.73                0.07                        0.0002   

   Half Instrument Uncertainty Squared  Total Uncertainty  
0                              0.00002               0.03  
1                              0.00002               0.02  
2                              0.00002               0.02  
3                              0.00002               0.02  
4                              0.00002               0.02

- **Conclusions from the statistical table:** As the sample size increases, both the mean and the squared standard uncertainty and the total uncertainty tend to decrease. On the other hand, the sample standard deviation increases. However, the changes in these measures are very small. Still, all values tend to approach the estimated measurement with total uncertainty mentioned previously, which was $T = (1.73 \pm 0.02) \space \text{seconds}$.

## Determinantion of G

3. Determination of g:

a) Use the direct measurements of l and T to calculate the value of the gravitational acceleration in the laboratory. Report your result by performing the uncertainty analysis using two methods:

- Elementary uncertainty theory

Using equation (2) to calculate gravity with the pendulum length and its period calculated in the previous point, that is, the most probable values of the pendulum length and period.

In [8]:
# Define the values to calculate g

pendulum_length = 0.736  # Length in meters
pendulum_period = 1.73   # Period value calculated in the previous point

# Calculating gravitational acceleration

calculated_gravity = (4 * (np.pi ** 2) * pendulum_length) / pendulum_period ** 2

print(f'Gravity = {round(calculated_gravity, 2)} meters/second^2')

Gravity = 9.71 meters/second^2


Uncertainty by analyzing extreme values

In [9]:
min_gravity = (4 * (np.pi ** 2) * pendulum_length) / (average_periods.max().max() ** 2)
max_gravity = (4 * (np.pi ** 2) * pendulum_length) / (average_periods.min().min() ** 2)

elementary_theory_uncertainty = round((max_gravity - min_gravity) / 2, 2)
print(f'Uncertainty by elementary theory analyzing extreme values: {elementary_theory_uncertainty}')

Uncertainty by elementary theory analyzing extreme values: 1.04


Thus, the gravitational acceleration by elementary uncertainty theory would be:
$$ g = (9.71 \pm 1.02)\space \frac{m}{s^2} $$

- Usando propagación por derivadas parciales de incertidumbres pequeñas.

$$\Delta g = \frac{\partial g}{\partial l}\Big |_l \cdot \Delta l + \frac{\partial g}{\partial T} \Big |_T \cdot \Delta T$$

Thus: $$\Delta g = \frac{4\pi^2}{(1.73 s)^2}\cdot(0.001 m) - \frac{2\pi^2(0.736 m)}{(1.73 s)} \cdot(0.02s) = 0.15 \frac{m}{s^2}$$

This error calculated by partial derivatives is much smaller than the one obtained with the previous method using elementary uncertainty theory. This is because during the measurement process there were some errors that caused the extreme value analysis in the previous section to yield very large values due to these significant differences. In conclusion, the partial derivatives method is more precise since it does not take into account procedural errors as significant as the outliers in the measurements.

b) Could we say that one of the variables presents a dominant uncertainty compared to the other in the context of uncertainty analysis and propagation?

The influence of the time factor generates a significantly higher uncertainty compared to that coming from the length, because the magnitude of the expression $ \frac{\partial g}{\partial T} \Big |_T \cdot \Delta T$ is significantly greater.

c) Calculating the gravitational acceleration in Medellín and the relative error with respect to the value obtained in the pendulum experiment, as well as comparing it to the value given by (Estaciones Red de Gravedad Absoluta Para Colombia -RGAC-, 2024).

In [10]:
latitude = radians(6.2)
altitude_above_sea = 1464
A, B = 0.0053024, 0.0000058

expected_gravity = 9.780327 * (1 + (A * latitude) - (B * 2 * latitude)) - (3.086e-6 * altitude_above_sea)
rgac_gravity = 9.77741092 # (m/s^2)

relative_error_calculated = (abs(expected_gravity - calculated_gravity) / expected_gravity) * 100
relative_error_rgac = (abs(expected_gravity - rgac_gravity) / expected_gravity) * 100
print(f'The expected gravity value is: {round(expected_gravity, 4)} meters/second^2')
print(f'Relative error between the expected and calculated measurement: {round(relative_error_calculated, 2)}%')

print(f'\nThe RGAC gravity value is: {round(rgac_gravity, 4)} meters/second^2')
print(f'Relative error between the expected and RGAC value: {round(relative_error_rgac, 2)}%')

The expected gravity value is: 9.7814 meters/second^2
Relative error between the expected and calculated measurement: 0.75%

The RGAC gravity value is: 9.7774 meters/second^2
Relative error between the expected and RGAC value: 0.04%


# Results

The experimental measurements of the pendulum oscillation period gave consistent results across all trials, the oscillation times were recorded for 1, 5, 7, 9, and 10 oscillations, and by dividing each measurement by the corresponding number of oscillations, the individual period values were obtained. The average period values for each oscillation set were: 1.60 s (1 oscillation), 1.76 s (5 oscillations), 1.75 s (7 oscillations), 1.77 s (9 oscillations), and 1.76 s (10 oscillations). After combining all 25 period measurements and performing statistical analysis, the best estimate of the pendulum period was determined to be $T = (1.73 \pm 0.02)$ seconds, where the total uncertainty accounts for both the standard uncertainty of the mean and the stopwatch resolution (0.01 s). The statistical table analysis demonstrated that as the sample size increases from 5 to 30 measurements, the mean stabilizes around 1.73 s while the total uncertainty decreases, confirming the benefit of repeated measurements. Using the measured pendulum length of $l = (73.6 \pm 0.1)$ cm and the calculated period, the gravitational acceleration was determined to be $g = 9.71 \, \text{m/s}^2$. 

Two methods were employed for uncertainty propagation: the elementary uncertainty theory using extreme values gave $g = (9.71 \pm 1.02) \, \text{m/s}^2$, while the partial derivatives method gave a much smaller uncertainty of $\Delta g = 0.15 \, \text{m/s}^2$, resulting in $g = (9.71 \pm 0.15) \, \text{m/s}^2$. The expected gravity value for Medellín, calculated using the latitude (6.2°) and altitude (1464 m above sea level), was found to be approximately 9.78 m/s². The relative error between the experimentally calculated gravity and the expected value was approximately 0.7%, while the RGAC reference value of 9.77741092 m/s² showed an even smaller discrepancy with the theoretical prediction.

# Conclusions

This experiment successfully demonstrated the application of uncertainty analysis techniques in the determination of gravitational acceleration using a simple pendulum. The results confirm that repeated measurements are essential for obtaining reliable estimates of physical quantities, as they allow the calculation of statistical measures such as the mean, standard deviation, and standard uncertainty, which together provide a comprehensive picture of measurement precision. The analysis revealed that the uncertainty associated with the period measurement dominates the overall uncertainty in the calculated gravitational acceleration, as shown by the partial derivatives propagation method where the contribution from $\frac{\partial g}{\partial T} \cdot \Delta T$ is significantly larger than that from $\frac{\partial g}{\partial l} \cdot \Delta l$. This finding underscores the importance of minimizing timing errors in pendulum experiments, which can be achieved by measuring multiple oscillations and dividing by the number of cycles, as well as by maintaining consistent experimental conditions. The comparison between the two uncertainty propagation methods highlighted that the elementary theory using extreme values tends to overestimate the uncertainty when outliers are present in the data, whereas the partial derivatives method provides a more refined estimate by considering only small, random fluctuations around the best estimate values. The experimentally obtained gravitational acceleration of $g = 9.71 \, \text{m/s}^2$ showed excellent agreement with both the theoretical prediction for Medellín and the high-precision measurement from the Colombian Absolute Gravity Network (RGAC), with a relative error of less than 1%. This close agreement validates the experimental methodology and confirms that even with simple laboratory equipment, meaningful and accurate physical measurements can be achieved when proper uncertainty analysis is applied. Future improvements could include taking multiple measurements of the pendulum length and using electronic timing systems to further reduce human reaction time errors.

# References
- Estaciones Red de Gravedad Absoluta para Colombia -RGAC-. (2024). Sgc.gov.co. https://datos.sgc.gov.co/datasets/b17d33bba0c444fbb7899ae1ee26e5f2_0/explore?location=6.200292%2C-75.578486%2C17.90

- Notas de Física experimental — Notas de Física Experimental. (2022). Srco.org.co. https://www.srco.org.co/NotasFExp/intro.html#

- ‌Mahecha, Jorge. (2009). TratamientoEstadísticoDeDatosExperimentales. https://www.researchgate.net/publication/388003805_TratamientoEstadisticoDeDatosExperimentales